quinn-udp/src/windows.rs

```diff
@@ -68,38 +133,162 @@ impl UdpSocketState {
+                    let ip = std::net::SocketAddr::new(ip, 0);
+                    let ip = socket2::SockAddr::from(ip);
+                    match ip.family() {
+                        WinSock::AF_INET => {
+                            let src_ip: WinSock::SOCKADDR_IN =
+                                unsafe { ptr::read(ip.as_ptr() as _) };
+                            let pktinfo = WinSock::IN_PKTINFO {
+                                ipi_addr: src_ip.sin_addr,
+                                ipi_ifindex: 0,
+                            };
+                            encoder.push(WinSock::IPPROTO_IP, WinSock::IP_PKTINFO, pktinfo);
+                        }
+                        WinSock::AF_INET6 => {
+                            let src_ip: WinSock::SOCKADDR_IN6 =
+                                unsafe { ptr::read(ip.as_ptr() as _) };
+                            let pktinfo = WinSock::IN6_PKTINFO {
+                                ipi6_addr: src_ip.sin6_addr,
+                                ipi6_ifindex: unsafe { src_ip.Anonymous.sin6_scope_id },
+                            };
+                            encoder.push(WinSock::IPPROTO_IPV6, WinSock::IPV6_PKTINFO, pktinfo);
+                        }
+                        _ => {
+                            return Err(io::Error::from(io::ErrorKind::InvalidInput));
+                        }
                     }
+                }
+
+                encoder.finish();
+
+                let mut len = 0;
+                let rc = unsafe {
+                    WinSock::WSASendMsg(
+                        socket.0.as_raw_socket() as usize,
+                        &wsa_msg,
+                        0,
+                        &mut len,
+                        ptr::null_mut(),
+                        None,
+                    )
+                };
 
+                if rc == 0 {
+                    sent += 1;
+                } else if sent != 0 {
+                    // We need to report that some packets were sent in this case, so we rely on
+                    // errors being either harmlessly transient (in the case of WouldBlock) or
+                    // recurring on the next call.
+                    return Ok(sent);
+                } else if rc == WinSock::WSAEWOULDBLOCK {
+                    return Err(io::Error::last_os_error());
+                } else {
                     // Other errors are ignored, since they will usually be handled
                     // by higher level retransmits and timeouts.
-                    log_sendmsg_error(&self.last_send_error, e, transmit);
+                    log_sendmsg_error(&self.last_send_error, io::Error::last_os_error(), transmit);
                     sent += 1;
                 }
+            } else {
+                match socket.0.send_to(
+                    &transmit.contents,
+                    &socket2::SockAddr::from(transmit.destination),
+                ) {
```

Comment:
Let's not duplicate the body of this `match`. Maybe compute a `Result` from each possible send path, and then match on it once?

Reply:
ok i will do it